api/src/services/survey-block-service.ts

```diff
@@ -0,0 +1,89 @@
+import { IDBConnection } from '../database/db';
+import { SurveyBlock, SurveyBlockRecord, SurveyBlockRepository } from '../repositories/survey-block-repository';
+import { DBService } from './db-service';
+
+export class SurveyBlockService extends DBService {
+  surveyBlockRepository: SurveyBlockRepository;
+
+  constructor(connection: IDBConnection) {
+    super(connection);
+    this.surveyBlockRepository = new SurveyBlockRepository(connection);
+  }
+
+  /**
+   * Gets Block Survey Records for a given survey id
+   *
+   * @param {number} surveyId
+   * @return {*} {Promise<SurveyBlockRecord[]>}
+   * @returns
+   */
+  async getSurveyBlocksForSurveyId(surveyId: number): Promise<SurveyBlockRecord[]> {
+    return await this.surveyBlockRepository.getSurveyBlocksForSurveyId(surveyId);
+  }
+
+  /**
+   *  Deletes a survey block record.
+   *
+   * @param {number} surveyBlockId
+   * @return {*}  {Promise<SurveyBlockRecord>}
+   * @memberof SurveyBlockService
+   */
+  async deleteSurveyBlock(surveyBlockId: number): Promise<SurveyBlockRecord> {
+    return this.surveyBlockRepository.deleteSurveyBlockRecord(surveyBlockId);
+  }
+
+  /**
+   * Insert or Updates Survey Block Records based on the existence of a survey_block_id
+   *
+   * @param {number} surveyId
+   * @param {SurveyBlock[]} blocks
+   * @return {*} {Promise<void>}
+   * @memberof SurveyBlockService
+   */
+  async updateInsertSurveyBlocks(surveyId: number, blocks: SurveyBlock[]): Promise<void> {
+    const insertUpdate: Promise<any>[] = [];
+
+    blocks.forEach((item: SurveyBlock) => {
+      item.survey_id = surveyId;
+      if (item.survey_block_id) {
+        insertUpdate.push(this.surveyBlockRepository.updateSurveyBlock(item));
+      } else {
+        insertUpdate.push(this.surveyBlockRepository.insertSurveyBlock(item));
+      }
+    });
+
+    await Promise.all(insertUpdate);
+  }
+
+  /**
+   * Inserts, Updates and Deletes Block records
+   * All passed in blocks are treated as the source of truth,
+   * Any pre existing blocks that do not collide with passed in blocks are deleted
+   *
+   * @param {number} surveyId
+   * @param {SurveyBlock[]} blocks
+   * @return {*} {Promise<void>}
+   * @memberof SurveyBlockService
+   */
+  async upsertSurveyBlocks(surveyId: number, blocks: SurveyBlock[]): Promise<void> {
```

Comment:
I find it a little confusing that there are two methods with very siimilar names. Because they take the same parameters, this would probably cause confusion in the future. My instinct would be to delete the `updateInsertSurveyBlocks` method and move its logic into this one. But it's your call!

Reply:
Overall I think your approach to the problem is sound 👍

Reply:
I agree it is confusing.

Reply:
combined the functions